File: app/src/main/java/com/projects/johnny/myway/AddLocationFragment.java
package com.projects.johnny.myway;

import android.animation.Animator;
import android.app.Fragment;
import android.app.FragmentManager;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.view.inputmethod.InputMethod;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.FrameLayout;

import com.firebase.client.Firebase;
import com.google.android.gms.common.api.Status;
import com.google.android.gms.location.places.Place;
import com.google.android.gms.location.places.ui.PlaceAutocompleteFragment;
import com.google.android.gms.location.places.ui.PlaceSelectionListener;

import io.codetail.animation.ViewAnimationUtils;
import io.codetail.widget.RevealFrameLayout;

public class AddLocationFragment extends Fragment {

    private static final String TAG = "AddLocationFragment";

    public static AddLocationFragment newInstance() {
        return new AddLocationFragment();
    }

    private String locationAddress = "";
    private EditText mNicknameEditText;

    private RevealFrameLayout mContainer;
    private FrameLayout mCircularRevealView;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_add_location, container, false);

        mNicknameEditText = (EditText) v.findViewById(R.id.add_location_nickname_edit_text);

        // Firebase
        App app = (App) getActivity().getApplicationContext();
        String UID = app.getUID();
        Firebase.setAndroidContext(getActivity().getApplicationContext());
        final Firebase mFirebaseRef = new Firebase("https://myways.firebaseIO.com/").child(UID).child("Locations");

        // Get reference to button and have it initially disabled
        final Button mConfirmAddPlaceButton = (Button) v.findViewById(R.id.confirm_add_place_button);
        mConfirmAddPlaceButton.setEnabled(false);

        // Because we are inside a fragment and we want to obtain a fragment,
        // we have to use the childFragmentManager.
        final FragmentManager fm = getChildFragmentManager();

        // Get reference to PlaceAutocomplete UI widget
        final PlaceAutocompleteFragment mPlaceAutocompleteFragment = (PlaceAutocompleteFragment) fm.findFragmentById(R.id.place_autocomplete_fragment);

        // Set listener for autocomplete widget
        mPlaceAutocompleteFragment.setOnPlaceSelectedListener(new PlaceSelectionListener() {

            @Override
            public void onPlaceSelected(Place place) {
                // Obtain address from place selected
                locationAddress = place.getAddress().toString();
                mPlaceAutocompleteFragment.setText(locationAddress);
                // Enable button after selecting a place
                mConfirmAddPlaceButton.setEnabled(true);
            }

            @Override
            public void onError(Status status) {
                Log.i(TAG, "An error occurred: " + status);
            }
        });

        // Set listener for button
        mConfirmAddPlaceButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // Hide keyboard
                InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
                imm.hideSoftInputFromWindow(mNicknameEditText.getWindowToken(), 0);

                // Add location to storage with or without nickname, depending
                //  on what the user wants.
                String nickname;
                if (mNicknameEditText.getText().length() > 0) {
                    nickname = mNicknameEditText.getText().toString();
                    mFirebaseRef.child(nickname).setValue(locationAddress);
                } else {
                    nickname = locationAddress;
                    mFirebaseRef.child(nickname).setValue(locationAddress);
                }

                returnToDirectionsFragment();
            }
        });

        mPlaceAutocompleteFragment.setHint("Search Places");

        // For UP navigation button
//        getActivity().getActionBar().setDisplayHomeAsUpEnabled(true);
//        getActivity().getActionBar().setDisplayShowHomeEnabled(true);

        // Animation stuff
        mContainer = (RevealFrameLayout) v.findViewById(R.id.add_location_container);
        mCircularRevealView = (FrameLayout) v.findViewById(R.id.add_location_circular_reveal_view);

        // Animate circular reveal
        mContainer.post(new Runnable() {
            @Override
            public void run() {
                // Determine center position for circular reveal (fab button)
                final int x = mContainer.getRight() / 2;
                final int y = mContainer.getBottom() / 2;

                // Determine radius sizes
                final int containerWidth = mContainer.getWidth();
                final int containerHeight = mContainer.getHeight();

                final float startingRadius = (float) Math.sqrt((containerWidth * containerWidth) + (containerHeight * containerHeight));;
                final float endRadius = 0;

                final Animator animator = ViewAnimationUtils.createCircularReveal(mCircularRevealView, x, y, startingRadius, endRadius);
                animator.setDuration(400);
                animator.addListener(new Animator.AnimatorListener() {
                    @Override
                    public void onAnimationStart(Animator animation) {

                    }

                    @Override
                    public void onAnimationEnd(Animator animation) {
                        mCircularRevealView.setVisibility(View.INVISIBLE);
                    }

                    @Override
                    public void onAnimationCancel(Animator animation) {

                    }

                    @Override
                    public void onAnimationRepeat(Animator animation) {

                    }
                });

                animator.start();
            }
        });

        return v;
    }

    private void returnToDirectionsFragment() {
        // Remember that we're using FragmentManager here rather than ChildFragmentManager
        final FragmentManager fm = getFragmentManager();
//        // Return to previous fragment
        fm.popBackStack();
    }
}

File: app/src/main/java/com/projects/johnny/myway/AddLocationFragment.kt
package com.projects.johnny.myway

import android.animation.Animator
import android.app.Fragment
import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import com.firebase.client.Firebase
import com.google.android.gms.common.api.Status
import com.google.android.gms.location.places.Place
import com.google.android.gms.location.places.ui.PlaceAutocompleteFragment
import com.google.android.gms.location.places.ui.PlaceSelectionListener
import io.codetail.animation.ViewAnimationUtils
import kotlinx.android.synthetic.main.fragment_add_location.*
import org.jetbrains.anko.enabled

class AddLocationFragment : Fragment() {

    companion object {
        fun newInstance() = AddLocationFragment()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View =
            inflater.inflate(R.layout.fragment_add_location, container, false)

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        // Firebase stuff
        val uid = App.getInstance().uid
        Firebase.setAndroidContext(activity.applicationContext)
        val firebaseRef = Firebase("https://myways.firebaseIO.com/").child(uid).child("Locations");
        addPlaceButton.enabled = false

        // Because we are INSIDE fragment and we are FINDING a fragment, we need to use childFragmentManager
        val autocompleteFrag = childFragmentManager.findFragmentById(R.id.placeAutocompleteFragment) as PlaceAutocompleteFragment
        var locationAddress: String? = null
        autocompleteFrag.setHint("Search Places")
        autocompleteFrag.setOnPlaceSelectedListener(object: PlaceSelectionListener {
            override fun onPlaceSelected(place: Place) {
                locationAddress = place.address.toString()
                autocompleteFrag.setText(locationAddress)
                // Enable utton after selecting a place
                // TODO: Gray out button when it is disabled?
                addPlaceButton.enabled = true
            }
            override fun onError(status: Status) {
                Log.e(tag, "An error occured: $status")
            }
        })

        addPlaceButton.setOnClickListener {
            // Hide keyboard
            val imm = activity.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            imm.hideSoftInputFromWindow(nickname.windowToken, 0)

            val nickname = if (nickname.text.length > 0) nickname.text.toString() else locationAddress
            firebaseRef.child(nickname).setValue(locationAddress)
            fragmentManager.popBackStack()
        }

        // Animation stuff
        addLocationContainer.post {
            with(addLocationContainer) {
                // Determine center position for circular reveal (fab button)
                // Radius sizes = width, height of addLocationContainer, respectively
                val x = right / 2
                val y = bottom / 2

                val startingRadius = Math.sqrt((width * width).toDouble() + (height * height).toDouble()).toFloat()
                val endRadius = 0.0f

                val animator = ViewAnimationUtils.createCircularReveal(circularRevealView, x, y, startingRadius, endRadius)
                animator.setDuration(400)
                animator.addListener(object: Animator.AnimatorListener {
                    override fun onAnimationRepeat(animation: Animator?) { }

                    override fun onAnimationEnd(animation: Animator?) {
                        circularRevealView.visibility = View.INVISIBLE
                    }

                    override fun onAnimationCancel(animation: Animator?) { }

                    override fun onAnimationStart(animation: Animator?) { }
                })
                animator.start()
            }
        }
    }
}